src/event.rs

```diff
@@ -347,49 +362,50 @@ where
 					}
 					PaymentPurpose::SpontaneousPayment(preimage) => (Some(preimage), None),
 				};
-				let mut payments = self.inbound_payments.lock().unwrap();
-				match payments.entry(payment_hash) {
-					hash_map::Entry::Occupied(mut e) => {
-						let payment = e.get_mut();
-						payment.status = PaymentStatus::Succeeded;
-						payment.preimage = payment_preimage;
-						payment.secret = payment_secret;
-						payment.amount_msat = Some(amount_msat);
-					}
-					hash_map::Entry::Vacant(e) => {
-						e.insert(PaymentInfo {
+
+				let payment_info =
+					if let Some(mut payment_info) = self.payment_store.get(&payment_hash) {
+						payment_info.status = PaymentStatus::Succeeded;
+						payment_info.preimage = payment_preimage;
+						payment_info.secret = payment_secret;
+						payment_info.amount_msat = Some(amount_msat);
+						payment_info
+					} else {
+						PaymentInfo {
 							preimage: payment_preimage,
+							payment_hash,
 							secret: payment_secret,
-							status: PaymentStatus::Succeeded,
 							amount_msat: Some(amount_msat),
-						});
-					}
-				}
+							direction: PaymentDirection::Inbound,
+							status: PaymentStatus::Succeeded,
+						}
+					};
+
+				self.payment_store.insert(payment_info).expect("Failed to access payment store");
```

Comment:
May want to expose something like `HashMap`'s `entry().and_modify().or_insert_with()` otherwise you'll lock twice and could race if events are processed in parallel. Perhaps locking shouldn't be internal to the store?